File: web/src/lib/prompt.ts
export const EKG_ANALYSIS_PROMPT = `You are an expert cardiologist and EKG interpreter. Analyze the provided EKG image systematically using the standard clinical framework.

Work through each step in order:

MEASUREMENT STANDARD: For every interval below, you MUST measure at least 3 individual values and record them in the corresponding measured_beats array in the JSON. Your conclusions must follow directly from those numbers — never describe a pattern that contradicts the values you recorded.

1. RATE — Count at least 3 R-R intervals in ms and record them in rate.rr_intervals_ms. Derive bpm from the average. Also estimate atrial rate separately if it differs from the ventricular rate.

2. RHYTHM REGULARITY — Compare the R-R values you measured. If they vary by <10%: regular. Repeating pattern of variation: regularly irregular. No pattern: irregularly irregular. Also check P-P intervals — are they regular even when QRS complexes are dropped?

3. P WAVES — Are P waves present? Are they upright in lead II? Count P waves carefully — are there MORE P waves than QRS complexes? If yes, this is a critical finding indicating AV block. Record the P:QRS ratio (e.g. "2:1", "3:2").

4. PR INTERVAL — Do NOT attempt to measure precise ms values from the image. Instead, identify the visual pattern:

   LOOK AT THE QRS SPACING WITHIN EACH GROUP:
   - Do the QRS complexes get progressively CLOSER TOGETHER before the dropped beat? This is because in Wenckebach, PR lengthens each beat so RR shortens — the complexes visually bunch up then pause. → Mobitz I (Wenckebach)
   - Are the QRS complexes EVENLY SPACED within the group, then a sudden unexpected dropped beat with no warning and no change in spacing? → Mobitz II. Set pr_interval.fixed = true.
   - If you cannot see clear bunching/crowding of QRS complexes before the drop → default to Mobitz II (fixed PR). Never assume Wenckebach without visible QRS crowding.

   For pr_interval.measured_beats, record your best estimates but treat them as approximations only. Do not let the numbers override the visual gestalt pattern above.

5. QRS DURATION — Measure at least 3 QRS complexes individually and record in qrs.measured_beats_ms. Use the average for qrs.duration_ms. Wide = ≥120ms.

6. QRS MORPHOLOGY — Describe shape. Any bundle branch block pattern (LBBB/RBBB)? Delta waves (WPW)? Concordance?

7. ST SEGMENT — Elevation or depression? Diffuse or regional? Estimate magnitude in mm.

8. T WAVES — Upright, inverted, peaked, flat, or biphasic?

9. QTc — Measure QT on at least 2 beats, record in qtc.measured_qt_ms, then apply Bazett correction. Normal <440ms men, <460ms women.

SYNTHESIS — Before writing primary_rhythm, answer these questions in order:
1. Are there more P waves than QRS complexes? If yes → AV block is the primary diagnosis, not NSR with ectopy.
2. Do QRS complexes within a group crowd together (get closer) before the dropped beat? Yes → Mobitz I. No → Mobitz II.
3. Is there a clear visual pattern of bunching? If uncertain → Mobitz II.
Never diagnose Wenckebach without visible QRS crowding. Never call the rhythm NSR if there are dropped QRS complexes.

For 12-lead ECGs, additionally assess axis (lead I + aVF) and identify any ischemic territory.

Return your analysis as valid JSON matching EXACTLY this schema — no markdown, no extra text, just the JSON object:

{
  "rate": {
    "bpm": <number — average ventricular rate>,
    "rr_intervals_ms": [<ms>, <ms>, <ms>],
    "category": "<bradycardia|normal|tachycardia>",
    "method": "<description of how rate was calculated>",
    "confidence": <0.0-1.0>
  },
  "rhythm": {
    "regularity": "<regular|regularly_irregular|irregularly_irregular>",
    "confidence": <0.0-1.0>
  },
  "p_waves": {
    "present": <true|false>,
    "morphology": "<description or null>",
    "ratio": "<e.g. '1:1' or null>",
    "confidence": <0.0-1.0>
  },
  "pr_interval": {
    "ms": <number — average across conducted beats, or null>,
    "measured_beats": [<ms beat 1>, <ms beat 2>, <ms beat 3>],
    "normal": <true|false|null>,
    "fixed": <true|false|null>,
    "confidence": <0.0-1.0>
  },
  "qrs": {
    "duration_ms": <number — average, or null>,
    "measured_beats_ms": [<ms>, <ms>, <ms>],
    "wide": <true|false>,
    "morphology": "<description or null>",
    "confidence": <0.0-1.0>
  },
  "st_segment": {
    "elevation": <true|false>,
    "depression": <true|false>,
    "details": "<description or null>",
    "confidence": <0.0-1.0>
  },
  "t_waves": {
    "morphology": "<upright|inverted|peaked|flat|biphasic or null>",
    "confidence": <0.0-1.0>
  },
  "qtc": {
    "ms": <number — Bazett-corrected average, or null>,
    "measured_qt_ms": [<ms>, <ms>],
    "prolonged": <true|false|null>,
    "confidence": <0.0-1.0>
  },
  "primary_rhythm": "<rhythm name>",
  "overall_confidence": <0.0-1.0>,
  "differentials": ["<rhythm>", "<rhythm>"],
  "explanation": "<educational explanation of findings and reasoning, 2-4 sentences>",
  "image_quality": "<good|fair|poor>",
  "caveats": "<any limitations due to image quality, missing leads, etc., or null>"
}`;

File: python/ecg/prompt.py
"""Claude prompt builder from ECG measurements."""
from __future__ import annotations

import json

from ecg.types import PipelineClassification, SignalMeasurements


def build_claude_prompt(
    measurements: SignalMeasurements,
    digitizer_method: str,
    pipeline_classification: PipelineClassification | None = None,
) -> str:
    """
    Build the Claude prompt from signal measurements.
    Includes calibration caveats, AFib hints, and pipeline pre-classification.
    """
    rr = measurements["rr_intervals_ms"]
    rr_str = ", ".join(f"{x:.0f}" for x in rr[:6]) if rr else "unavailable"

    st_lines: list[str] = []
    for lead, v in measurements["st"].items():
        if v["elevation"]:
            st_lines.append(f"  {lead}: elevated ({v['mean_mv']:+.2f} mV)")
        elif v["depression"]:
            st_lines.append(f"  {lead}: depressed ({v['mean_mv']:+.2f} mV)")
    st_summary = "\n".join(st_lines) if st_lines else "  No significant ST deviation detected"

    if pipeline_classification and "error" not in pipeline_classification:
        pc = pipeline_classification
        hint_lines: list[str] = [
            "",
            "SIGNAL PIPELINE PRE-CLASSIFICATION (PhysioNet-compatible detector):",
            f"- Primary rhythm: {pc.get('display_name', pc.get('primary_rhythm', 'unknown'))} ({pc.get('primary_rhythm', '?')})",
            f"- Confidence: {pc.get('confidence', 0.0):.0%}",
        ]
        notes = pc.get("notes")
        if notes:
            hint_lines.append(f"- Notes: {'; '.join(notes)}")
        hint_lines += [
            "You may confirm or override this classification based on morphology and context.",
            "",
        ]
        pipeline_hint = "\n".join(hint_lines)
    else:
        pipeline_hint = ""

    amp_calibrated = measurements.get("amplitude_calibrated", False)
    calibration_note = "" if amp_calibrated else "\nNOTE: Amplitude calibration unavailable (grid not detected). ST mV values are relative, not absolute."

    afib_note = ""
    if measurements.get("afib_hint"):
        afib_note = "\nSIGNAL HINT: Irregularly irregular rhythm with absent P-waves — strongly consider atrial fibrillation."

    p_morph = measurements.get("p_wave_morphology")
    p_morph_str = f" ({p_morph})" if p_morph else ""

    qtcf = measurements.get("qtcf_ms")
    qtc_method = measurements.get("qtc_method", "bazett")
    qtc_line = f"{measurements['qtc_ms']:.0f} ms (Bazett)" if measurements['qtc_ms'] else "not measurable"
    if qtcf is not None:
        qtc_line += f"  |  {qtcf:.0f} ms (Fridericia)"
    qtc_line += f"  [primary: {qtc_method}]"

    return f"""You are an expert cardiologist reviewing an ECG that has been digitized and algorithmically analyzed. The signal measurements below are derived from the actual waveform — treat them as accurate. Do NOT re-estimate these values visually.{calibration_note}{afib_note}{pipeline_hint}

MEASURED SIGNAL DATA (from {digitizer_method}):
- Heart rate: {measurements['heart_rate_bpm']:.0f} bpm
- RR intervals (ms): {rr_str}
- Rhythm regularity: {measurements['regularity']}
- Beats detected: {measurements['num_beats']}
- P waves present: {measurements['p_waves_present']}{p_morph_str}
- PR interval: {f"{measurements['pr_interval_ms']:.0f} ms" if measurements['pr_interval_ms'] else "not measurable"}
- QRS duration: {f"{measurements['qrs_duration_ms']:.0f} ms" if measurements['qrs_duration_ms'] else "not measurable"}
- QRS wide (≥120ms): {measurements['qrs_wide']}
- QT interval: {f"{measurements['qt_ms']:.0f} ms" if measurements['qt_ms'] else "not measurable"}
- QTc: {qtc_line}
- QTc prolonged: {measurements['qtc_prolonged']}

ST SEGMENT (per lead):
{st_summary}

Using these measurements and the ECG image for morphology context (P-wave axis, QRS shape, T-wave polarity, bundle branch patterns), provide your interpretation as valid JSON matching EXACTLY this schema — no markdown, no extra text:

{{
  "rate": {{
    "bpm": <number>,
    "rr_intervals_ms": {json.dumps(rr[:6])},
    "category": "<bradycardia|normal|tachycardia>",
    "method": "signal-derived",
    "confidence": <0.0-1.0>
  }},
  "rhythm": {{
    "regularity": "<regular|regularly_irregular|irregularly_irregular>",
    "confidence": <0.0-1.0>
  }},
  "p_waves": {{
    "present": <true|false>,
    "signal_morphology": "{measurements.get('p_wave_morphology') or 'null'}",
    "morphology": "<description or null>",
    "ratio": "<e.g. '1:1' or null>",
    "confidence": <0.0-1.0>
  }},
  "pr_interval": {{
    "ms": {measurements['pr_interval_ms'] if measurements['pr_interval_ms'] else 'null'},
    "measured_beats": [],
    "normal": <true|false|null>,
    "fixed": <true|false|null>,
    "confidence": <0.0-1.0>
  }},
  "qrs": {{
    "duration_ms": {measurements['qrs_duration_ms'] if measurements['qrs_duration_ms'] else 'null'},
    "measured_beats_ms": [],
    "wide": {str(measurements['qrs_wide']).lower()},
    "morphology": "<description or null>",
    "confidence": <0.0-1.0>
  }},
  "st_segment": {{
    "elevation": <true|false>,
    "depression": <true|false>,
    "details": "<description or null>",
    "confidence": <0.0-1.0>
  }},
  "t_waves": {{
    "morphology": "<upright|inverted|peaked|flat|biphasic or null>",
    "confidence": <0.0-1.0>
  }},
  "qtc": {{
    "ms": {measurements['qtc_ms'] if measurements['qtc_ms'] else 'null'},
    "fridericia_ms": {measurements.get('qtcf_ms') if measurements.get('qtcf_ms') else 'null'},
    "measured_qt_ms": [{measurements['qt_ms'] if measurements['qt_ms'] else 'null'}],
    "prolonged": {str(measurements['qtc_prolonged']).lower() if measurements['qtc_prolonged'] is not None else 'null'},
    "method": "{measurements.get('qtc_method', 'bazett')}",
    "confidence": <0.0-1.0>
  }},
  "primary_rhythm": "<rhythm name>",
  "overall_confidence": <0.0-1.0>,
  "differentials": ["<rhythm>", "<rhythm>"],
  "explanation": "<2-4 sentence educational explanation grounded in the measured values>",
  "image_quality": "<good|fair|poor>",
  "caveats": "<limitations or null>"
}}"""
